File: android/src/main/java/com/audiocontext/nodes/StereoPannerNode.kt
package com.audiocontext.nodes

import android.os.Build
import androidx.annotation.RequiresApi
import com.audiocontext.context.BaseAudioContext
import com.audiocontext.nodes.audionode.AudioNode
import com.audiocontext.nodes.audionode.ChannelCountMode
import com.audiocontext.parameters.AudioParam
import com.audiocontext.parameters.PlaybackParameters
import com.audiocontext.utils.Constants
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.sin

// https://webaudio.github.io/web-audio-api/#stereopanner-algorithm

class StereoPannerNode(
  context: BaseAudioContext,
) : AudioNode(context) {
  override val numberOfInputs: Int = 1
  override val numberOfOutputs: Int = 1
  override var channelCount: Int = 2
  override val channelCountMode: ChannelCountMode = ChannelCountMode.CLAMPED_MAX

  private val pan: AudioParam = AudioParam(context, 0.0, Constants.MAX_PAN, -Constants.MAX_PAN)
    get() = field

  @RequiresApi(Build.VERSION_CODES.N)
  override fun process(playbackParameters: PlaybackParameters) {
    mixBuffers(playbackParameters)

    val pan = pan.getValueAtTime(context.getCurrentTime())
    val x = (if (pan <= 0) pan + 1 else pan) * PI / 2

    val gainL = cos(x * PI / 2)
    val gainR = sin(x * PI / 2)

    // TODO: This assumes both channels play the same sound (buffer)
    // but it should mix the channels. F.e. pan < 0: leftChannel = leftChannel + leftGain * rightChannel

    if (pan <= 0) {
      playbackParameters.leftPan *= (1 + gainL)
      playbackParameters.rightPan *= gainR
    } else {
      playbackParameters.leftPan *= gainL
      playbackParameters.rightPan *= (1 + gainR)
    }

    super.process(playbackParameters)
  }
}

File: ios/nodes/StereoPannerNode/StereoPannerNode.m
#import "StereoPannerNode.h"
#import "AudioContext.h"

// https://webaudio.github.io/web-audio-api/#stereopanner-algorithm

@implementation StereoPannerNode

- (instancetype)initWithContext:(AudioContext *)context
{
  if (self = [super initWithContext:context]) {
    _panParam = [[AudioParam alloc] initWithContext:context value:0 minValue:-1 maxValue:1];
    self.numberOfInputs = 1;
    self.numberOfOutputs = 1;
  }

  return self;
}

- (void)cleanup
{
  _panParam = nil;
}

- (void)process:(AVAudioFrameCount)frameCount bufferList:(AudioBufferList *)bufferList;
{
  float time = [self.context getCurrentTime];
  float deltaTime = 1 / self.context.sampleRate;

  float *bufferL = (float *)bufferList->mBuffers[0].mData;
  float *bufferR = (float *)bufferList->mBuffers[1].mData;

  for (int frame = 0; frame < frameCount; frame += 1) {
    double pan = [_panParam getValueAtTime:[self.context getCurrentTime]];
    double x = ((pan <= 0) ? pan + 1 : pan) * M_PI / 2;

    double gainL = cos(x);
    double gainR = sin(x);

    if (pan <= 0) {
      bufferL[frame] += bufferR[frame] * gainL;
      bufferR[frame] *= gainR;
    } else {
      bufferL[frame] *= gainL;
      bufferR[frame] += bufferL * gainR;
    }

    time += deltaTime;
  }

  [super process:frameCount bufferList:bufferList];
};

@end
